quickevent/app/quickevent/plugins/CardReader/src/cardreaderplugin.cpp

```diff
@@ -298,6 +298,18 @@ int CardReaderPlugin::savePunchRecordToSql(const quickevent::core::si::PunchReco
 
 	int code = resolveAltCode(punch.code(), punch.stageid());
 
+	// check if punch isn't duplicate of existing saved punch
```

Comment:
what is this good for?

Reply:
It checks for duplicity in punches table.
Needed in case of : 
* new read `rawsplits` file (for example after stop/start service)
* second card redout with same finish time

Reply:
You are removing duplicates in `readAndProcessFile()` already, I like to have it all on one place. 

I do not know, why duplicate record in DB mind, but if you need to not-save-duplicate-punches do DB, please add comment with explanation before implementation, in other case, we do not know, what is this code good for. The best would be, if you can do this check in `readAndProcessFile()`. If this is not possible, please do not prepare sql query for such simple task.

Reply:
> You are removing duplicates in `readAndProcessFile()` already, I like to have it all on one place.

In `readAndProcessFile()` it remove duplicate from rawsplits file.
In `savePunchRecordToSql(...)` it checks if record doesn't exist in DB (finish from radio x finish from readout / 2nd readout of card)

> I do not know, why duplicate record in DB mind ...

The reason is, we don't want have duplicity in `punches` table. Because of Speaker module, which uses this table. If we don't check for duplicity on inserting in db, we need to check it on every shown in module Speaker.
I think it is better to resolve duplicity on write, and not on every read.

> ... please add comment with explanation before implementation, in other case, we do not know, what is this code good for. ...

I was hoping comment `check if punch isn't duplicate of existing saved punch` is enough. What information are you missing in this comment? I'm not sure what needs to be added?

Reply:
> If this is not possible, please do not prepare sql query for such simple task.

I was hoping it is more readable with `prepare`.  Can be rewrite to `q.exec("SELECT * ...`

Reply:
> > ... please add comment with explanation before implementation, in other case, we do not know, what is this code good for. ...
> 
> I was hoping comment `check if punch isn't duplicate of existing saved punch` is enough. What information are you missing in this comment? I'm not sure what needs to be added?

somethink like:
```
The reason is, we don't want have duplicity in punches table. Because of Speaker module, which uses this table. If we don't check for duplicity on inserting in db, we need to check it on every shown in module Speaker.
I think it is better to resolve duplicity on write, and not on every read.
```

Reply:
> > If this is not possible, please do not prepare sql query for such simple task.
> 
> I was hoping it is more readable with `prepare`. Can be rewrite to `q.exec("SELECT * ...`

Prepare query is more resource consuming for SQL server, they are useful in case, when one is executing the same query in the loop with different parameters. https://www.postgresql.org/docs/current/sql-prepare.html We are sometimes using prepared queries because `QSqlQuery::bind()` can correctly send QDateTime parameter to server with all the TZ stuff solved.

Reply:
> The reason is, we don't want have duplicity in punches table. Because of Speaker module, which uses this table. If we don't check for duplicity on inserting in db, we need to check it on every shown in module Speaker.
I think it is better to resolve duplicity on write, and not on every read.

This is not always true. 
1. If you want to state, that punch records shall be unique, then SQL unique index is the right way. Than you can have checks on many places or you can catch the SQL exceptions, but you do not have to think, if you would might to forget about duplicity check in some program path.
2. If you allow duplicates, then the better place to fix them is the place which do not need them. So `SELECT DISTINCT` is your friend.

This is something what must be discussed.

Reply:
I'll extend the comment & remove q.prepare.

>This is something what must be discussed.
I was hoping it is good enough. I'm not big expert in SQL. So, I'll wait until you resolve what is the best way to do it in SQL db. 
I just thought that duplicate data in the db are unwanted. They just increase the amount of data in the table and don't carry any information.